### PR TITLE
INT-9047 Make Summary.json available (to TAM or users)

### PR DESCRIPTION
**Pull request by eyadmba**
The overall feature is better explained here: https://github.com/JupiterOne/callisto/pull/418

The change done here allows a function to be called to do something with the results of an integration job.
That's needed for this feature so that the managed runtime can use that callback function to store the job's output in S3 (https://github.com/JupiterOne/jupiter-managed-integration/pull/353).

---

Jira Ticket: [INT-9047 Make Summary.json available (to TAM or users)
](https://jupiterone.atlassian.net/browse/INT-9047)

### CODE REVIEW

File: packages/integration-sdk-runtime/src/execution/executeIntegration.ts [hunk 1]
```diff
@@ -62,13 +62,14 @@ export interface ExecuteIntegrationOptions {
...
 }
 
-export interface ExecuteWithContextOptions {
-  graphObjectStore?: GraphObjectStore;
-  createStepGraphObjectDataUploader?: CreateStepGraphObjectDataUploaderFunction;
-}
+type ExecuteWithContextOptions = Pick<
+  ExecuteIntegrationOptions,
+  'graphObjectStore' | 'resultsCallback' | 'createStepGraphObjectDataUploader'
+>;
```
Comment (eyadmba) [inline packages/integration-sdk-runtime/src/execution/executeIntegration.ts:72]: Seems less error-prone than duplicating the interface.

File: packages/integration-sdk-runtime/src/execution/executeIntegration.ts [hunk 2]
```diff
@@ -288,6 +290,17 @@ export async function executeWithContext<
         data: summary,
       });
 
+      if (resultsCallback != null) {
+        try {
+          await resultsCallback(integrationStepResults);
+        } catch (err) {
+          context.logger.warn(
```
Comment (eyadmba) [inline packages/integration-sdk-runtime/src/execution/executeIntegration.ts:297]: Do we have a convention for appropriate log levels?

File: packages/integration-sdk-runtime/src/execution/executeIntegration.ts [hunk 3]
```diff
@@ -288,6 +290,17 @@ export async function executeWithContext<
         data: summary,
       });
 
+      if (resultsCallback != null) {
+        try {
+          await resultsCallback(integrationStepResults);
```
Comment (dobregons) [inline packages/integration-sdk-runtime/src/execution/executeIntegration.ts:295]: The summary.json that we store locally is something like this:
```typescript
{
  integrationStepResults,
  metadata: {
    partialDatasets,
  },
}
```
I wonder if we should add the `metadata` object, same as the existing version?

Comment (eyadmba) [reply packages/integration-sdk-runtime/src/execution/executeIntegration.ts:295]: Maybe I was over-engineering, but I thought the `resultsCallback` should provide the raw results, and then the function is responsible for summarizing it or whatever.

But if we don't think that's useful I can just pass in the `summary` instead.
Or I can also have the function that stores in S3 in the managed runtime run the summarization function before it sends it to S3.

Comment (dobregons) [reply packages/integration-sdk-runtime/src/execution/executeIntegration.ts:295]: I think the current version is fine, if we need to change it we can do it later, but IMO it's fine!

**Review by dobregons — APPROVED** [review]

**Pull request merged**